wlm/src/main/java/ai/djl/serving/wlm/Adapter.java

```diff
@@ -0,0 +1,172 @@
+/*
+ * Copyright 2023 Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"). You may not use this file except in compliance
+ * with the License. A copy of the License is located at
+ *
+ * http://aws.amazon.com/apache2.0/
+ *
+ * or in the "license" file accompanying this file. This file is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES
+ * OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+package ai.djl.serving.wlm;
+
+import ai.djl.inference.Predictor;
+import ai.djl.serving.wlm.WorkerPoolConfig.ThreadConfig;
+import ai.djl.serving.wlm.util.WorkerJob;
+
+import java.util.concurrent.CompletableFuture;
+
+/**
+ * An adapter is a modification producing a variation of a model that can be used during prediction.
+ */
+public abstract class Adapter {
+
+    protected String name;
+    protected String src;
+
+    /**
+     * Constructs an {@link Adapter}.
+     *
+     * @param name the adapter name
+     * @param src the adapter source
+     */
+    protected Adapter(String name, String src) {
+        this.name = name;
+        this.src = src;
+    }
+
+    /**
+     * Constructs a new {@link Adapter}.
+     *
+     * <p>After registration, you should call {@link #register(WorkerPool)}. This doesn't affect the
+     * worker pool itself.
+     *
+     * @param wp the worker pool for the new adapter
+     * @param name the adapter name
+     * @param src the adapter source
+     * @return the new adapter
+     */
+    public static Adapter newInstance(WorkerPool<?, ?> wp, String name, String src) {
+        if (!(wp.getWpc() instanceof ModelInfo)) {
+            String modelName = wp.getWpc().getId();
+            throw new IllegalArgumentException("The worker " + modelName + " is not a model");
+        }
+        ModelInfo<?, ?> modelInfo = (ModelInfo<?, ?>) wp.getWpc();
+        String engineName = modelInfo.getEngineName();
+        if ("Python".equals(engineName)) {
+            return new PyAdapter(name, src);
+        } else {
+            throw new IllegalArgumentException(
+                    "Adapters are only currently supported for Python models");
+        }
+    }
+
+    /**
+     * Unregisters an adapter in a worker pool.
+     *
+     * <p>This unregisters it in the wpc for new threads and all existing threads.
+     *
+     * @param wp the worker pool to remove the adapter from
+     * @param adapterName the adapter name
+     * @param <I> the input type
+     * @param <O> the output type
+     */
+    public static <I, O> void unregister(WorkerPool<I, O> wp, String adapterName) {
+        ModelInfo<I, O> wpc = (ModelInfo<I, O>) wp.getWpc();
+        Adapter adapter = wpc.unregisterAdapter(adapterName);
+        for (WorkerGroup<I, O> wg : wp.getWorkerGroups().values()) {
```

Comment:
maybe add a TODO to allow model-instance level adapter scheduling instead of register/unregister all.